src/parser/symtab.ts

```diff
@@ -0,0 +1,1738 @@
+        this.__module_scope = module_scope;
+    }
+
+    get [Symbol.toStringTag]() {
+        return '<symbol "' + this.__name + '">';
+    }
+
+    get_name() {
+        return this.__name;
+    }
+
+    is_referenced() {
+        return !!(this.__flags & SYMTAB_CONSTS.USE);
+    }
+
+    is_parameter() {
+        return !!(this.__flags & SYMTAB_CONSTS.DEF_PARAM);
+    }
+
+    is_global() {
+        /* Return *True* if the symbol is global. */
+        return !!(
+            this.__scope in [SYMTAB_CONSTS.GLOBAL_IMPLICIT, SYMTAB_CONSTS.GLOBAL_EXPLICIT] ||
+            (this.__module_scope && !!(this.__flags & SYMTAB_CONSTS.DEF_BOUND))
+        );
+    }
+
+    is_nonlocal() {
+        return !!(this.__flags & SYMTAB_CONSTS.DEF_NONLOCAL);
+    }
+
+    is_declared_global() {
+        return !!(this.__scope == SYMTAB_CONSTS.GLOBAL_EXPLICIT);
+    }
+
+    is_local() {
+        /* Return *True* if the symbol is local. */
+        return !!(
+            this.__scope in [SYMTAB_CONSTS.LOCAL, SYMTAB_CONSTS.CELL] ||
+            (this.__module_scope && this.__flags & SYMTAB_CONSTS.DEF_BOUND)
+        );
+    }
+
+    is_annotated() {
+        return !!(this.__flags & SYMTAB_CONSTS.DEF_ANNOT);
+    }
+
+    is_free() {
+        return !!(this.__scope == SYMTAB_CONSTS.FREE);
+    }
+
+    is_imported() {
+        return !!(this.__flags & SYMTAB_CONSTS.DEF_IMPORT);
+    }
+
+    is_assigned() {
+        return !!(this.__flags & SYMTAB_CONSTS.DEF_LOCAL);
+    }
+
+    is_comprehension() {
+        return !!(this.__flags & SYMTAB_CONSTS.DEF_COMP_ITER);
+    }
+
+    is_namespace() {
+        /*
+        Returns true if name binding introduces new namespace.
+
+        If the name is used as the target of a function or class
+        statement, this will be true.
+
+        Note that a single name can be bound to multiple objects.  If
+        is_namespace() is true, the name may also be bound to other
+        objects, like an int or list, that does not introduce a new
+        namespace.
+        */
+
+        return !!this.__namespaces;
+    }
+
+    get_namespaces() {
+        /* Return a list of namespaces bound to this name */
+        return this.__namespaces;
+    }
+
+    get_namespace() {
+        /*
+        Returns the single namespace bound to this name.
+
+        Raises ValueError if the name is bound to multiple namespaces.
+        */
+
+        if (this.__namespaces?.length != 1) {
+            throw Error("name is bound to multiple namespaces");
+            // todo: This should be a value error maybe
+        }
+
+        return this.__namespaces[0];
+    }
+}
+
+var astScopeCounter = 0;
+
+class SymbolTableScope {
+    name: string;
+    varnames: string[];
+    children: SymbolTableScope[];
+    blockType: BlockTypes;
+    isNested = false;
+    hasFree: boolean;
+    childHasFree: boolean;
+    generator: boolean;
+    varargs: boolean;
+    varkeywords: boolean;
+    returnsValue: boolean;
+    lineno: number;
```

Comment:
Are we missing `ste_needs_class_closure` equivalent? this should also be updated in `drop_class_free`
Not sure we have `drop_class_free`. But it gets called a bit later.

Reply:
I haven't reached `dropClassFree` yet.